Generics.md

```diff
@@ -1,164 +1,386 @@
+> (or one alternative way to see it, would be that all literal values are also subtypes, then '10' is both an instance of `AbstractInt` and a subtype of it. But it doesn't matter here.)
+
+## Formalism: Generic Type
+
+A generic type is associated with a type constraint. It can be replaced by any other type that is a subtype of the constraint.
+
+Said formally: Let `T` be a generic type and `C` its constraint (`T: C`). Then `U` can *specialize* `T` if `U: C`.
+
+A [generic declaration](#generic-declarations) (or built-in *type generator*) can be specialized with a generic type. That specialized type is also a generic type, e.g. `array<T>` is a generic type.
+
+### Operations on generic types
+
+In generic code, the following operations are possible:
+* instantiating a `T` with the zero-valued constructor,
+* calling a generic function by specializing its generic parameter `U: D` with `T` if `T: D`,
+* if `C` is a struct type, accessing members of `C` from an instance of `T`,
+* if `C` is an array or vector type, using the array indexing operator on an instance of `T`,
+* if `C` is a ptr type, using the dereference operator on an instance of `T`,
+* if `C` is a number type, using any built-in operations on numbers that are implemented for all subtypes of `C`,
+* ...etc.
+
+But the following is NOT possible:
+* assigning a `C` to a `T` (since `T` is a subtype of `C`, not necessarily equal)
+* calling a function that takes a parameter of type `C` with an instance of `T` (since `T` is a subtype of `C`, not necessarily equal)
+* ...etc.
+
+### Constraining generic types
+
+Note that compared to the WGSL spec, we need to constrain a bit more what can be a generic type. Consider this example from the spec:
+
+```rs
+@const @must_use fn array<T, N>(e1 : T, ..., eN : T) -> array<T, N>
+@must_use fn arrayLength(p: ptr<storage, array<E>, AM>) -> u32
+```
+`array` is parameterized by `T` and `N`. `arrayLength` is parameterized by `E` and `AM`.
+* `T` and `E` must be one of: a `scalar`, `vector`, `matrix`, `atomic`, `array`, structure[^1].
+* `N` must be an override-expression that evaluates to a `i32` or `u32`, and greater than 0[^2].
+* `AM` must be a value of enumerant [`access mode`](https://www.w3.org/TR/WGSL/#access-mode)[^3].
+
+So obviously, this level of constraining is too high to express in the grammar.
+
+TL;DR instead, we could specify these constraints:
+* Unconstrained (or implicitly constrained by [AllTypes](https://www.w3.org/TR/WGSL/#alltypes-type).
+* Constrained by: a user-defined struct, `vec`, `mat`, `array`, `atomic`, `ptr`, `AbstractInt` or `AbstractFloat`: Must be specialized by a *constructible* subtype.
+* Constrained by a leaf type: `u32`, `i32`, `f32`, `f16`, `bool`: Must be specialized with a instance of that type (like the `N` parameter).
+
+## Generic Declarations
+
+We introduce a `@generic` attribute to define the generic parameters and constraints.
+```rs
+@generic(Num: AbstractFloat, T: MyStruct, N: u32)
```

Comment:
Putting the generic paramaters between the fn name and the runtime parameters list is more traditional, no?
e.g. `fn foo<A: MyStruct, B>(a:A, b:B) { }`

Why not do that? Are you trying to minimize the new syntax? Does it create a parsing ambiguity?